### DISCUSSION TITLE
Server-side fetch with relative URLs

### DISCUSSION THREAD

**Discussion by dtinth** [remix-run/remix#5349] [Proposals]
It would be great if Remix supported relative URLs when doing a `fetch()` from the server-side, like SvelteKit.

In SvelteKit, you can [make a `fetch()` request on the server using a relative URL](https://kit.svelte.dev/docs/load#making-fetch-requests). It allows sending a request API routes on the same project, but **without the overhead of an actual HTTP call.**

In Remix,

```
TypeError: Invalid URL
    at new NodeError (node:internal/errors:393:5)
    at URL.onParseError (node:internal/url:565:9)
    at new URL (node:internal/url:645:5)
    at new Request (/Users/dtinth/ghq/github.com/creatorsgarten/contentsgarten/node_modules/.pnpm/@remix-run+web-fetch@4.3.2/node_modules/@remix-run/web-fetch/src/request.js:78:16)
    at /Users/dtinth/ghq/github.com/creatorsgarten/contentsgarten/node_modules/.pnpm/@remix-run+web-fetch@4.3.2/node_modules/@remix-run/web-fetch/src/fetch.js:42:19
    at new Promise (<anonymous>)
    at AsyncFunction.fetch (/Users/dtinth/ghq/github.com/creatorsgarten/contentsgarten/node_modules/.pnpm/@remix-run+web-fetch@4.3.2/node_modules/@remix-run/web-fetch/src/fetch.js:40:9)
    at fetch (/Users/dtinth/ghq/github.com/creatorsgarten/contentsgarten/node_modules/.pnpm/@remix-run+node@1.12.0/node_modules/@remix-run/node/dist/fetch.js:44:19)
```

## Why?

The project I’m building exposes both API routes and a UI routes. The API part accesses many 3rd party services (database, cloud providers, APIs, auth provider).

Ideally, the loaders in the UI routes will invoke 3rd party services through the API routes.

<img width="890" alt="image" src="https://user-images.githubusercontent.com/193136/216397907-fbe57378-de34-40f4-97a4-020a4f02c8d7.png">

Setting up local versions of these services can take a lot of work, especially if my task is to tweak some frontend. So I wanted to make it possible for developers to work on just the frontend part while connecting to a production backend.

Developers can choose to either run the front end against the same instance, or connect to the production instance (or any other instance).

<img width="1610" alt="image" src="https://user-images.githubusercontent.com/193136/216398480-8601d9d1-81d5-4e93-aba1-d85e3044a60a.png">

In SvelteKit, this is easy because a `fetch()` call accepts relative URL. However in Remix it’s not clear what the URL should be, especially when it’s deployed to a serverless provider.

I tried constructing an absolute URL from `request.url`, however on Netlify this results in the lambda function calling itself, which means it costs twice and there is double the overhead.

**Comment by sergiodxa** [comment]
I'm unconvinced this is something Remix should support, I understand the benefit but this goes against the standard of how `fetch` works, and if what you want is to share code between two routes the simplest way is to export a function

```ts
// app/services/get-items.ts
export async function getItems(args: GetItemsArgs): Promise<GetItemsResult> { ... }

// app/routes/api.items.ts
export async function loader({ request }: DataFunctionArgs) {
  await authorizeFromHeaders(request.headers);
  let argsFromRequest = extractArgsFromRequest()
  let items = await getItems(argsFromRequest)
  return json(items)
}

// app/routes/items.tsx
export async function loader({ request }: DataFunctionArgs) {
  await authorizeFromCookies(request);
  let argsFromRequest = extractArgsFromRequest()
  let items = await getItems(argsFromRequest)
  let anotherThing = await getAnotherThing(argsFromRequest)
  return json({
    items: reduceItemsDataToMinimum(items), // remove parts of items that UI doesn't need
    anotherThing
  })
}

export default function Component() {
  let { items, anotherThing } = useLoaderData<typeof loader>()
  // render ui using items and anotherThing
}
```

**Reply by dtinth** [reply]
> this goes against the standard of how fetch works

In browsers fetch **does** support a relative URL.

Both [WhatWG](https://fetch.spec.whatwg.org/#fetch-api) _and_ [WinterCG](https://fetch.spec.wintercg.org/#fetch-api) shows a lot of relative URLs being used:

> <img width="860" alt="image" src="https://user-images.githubusercontent.com/193136/216601310-54911bca-2ca2-4a6f-9aa3-edce62eb65cf.png">

Therefore I fail to see how supporting relative URLs “goes against the standard.” Can you elaborate more?

> if what you want is to share code between two routes the simplest way is to export a function

Yes, I want to share code, but not only that, I also want to make it possible for a `loader` to invoke a backend hosted in production.

- If `process.env.BACKEND_URL` is present, the loader should make a request to that backend.
- If `process.env.BACKEND_URL` is blank, the loader should load the data from API routes defined on the same server.

By using function calls, it would look like this:

```ts
  // Since I cannot use relative URLs here I have to handle these 2 cases separately.
  let items = process.env.BACKEND_URL
    ? await fetch(process.env.BACKEND_URL + '/api.items', { /* ... */ })
         .then(response => { /* ... handle errors, parse response ... */ }
    : await getItems(argsFromRequest)
```

But with relative URL support it will be much simpler:

```ts
  // When `BACKEND_URL` is blank we have a relative URL, and it will invoke the API on the dev server.
  // When `BACKEND_URL` is present, we have an absolute URL, and it will invoke the API on the remote server.
  let items = await fetch(`${process.env.BACKEND_URL || ''}/api.items`, { /* ... */ })
    .then(response => { /* ... handle errors, parse response ... */ }
```

**Reply by sergiodxa** [reply]
> this goes against the standard of how fetch works

Because as far as I understood you, your proposal and what SveltKit does is that if you call `fetch("/api/items")` instead of doing a network request it will do a function call to the loader of the route matching `/api/items`

Fetch does a network call, not a function call, if this is implemented it would change the semantics of what Fetch does when using it server-side in Remix.
